### PR TITLE
lib/vfscore: Fix behavior of writable open on directories

### PR DESCRIPTION
**Pull request by andreittr**
### Description of changes

Previously the open syscall, when called on a directory with write mode would either fail with EINVAL or, worse, finish successfully, whereas the expected behavior is to return EISDIR.
This change corrects this behavior, bringing open() in line with expectations.

### Prerequisite checklist

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.uk`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.uk) on your commit series before opening this PR;
 - [x] Updated relevant documentation.

### Base target

 - Architecture(s): all
 - Platform(s): all
 - Application(s): all

### Additional configuration

Minimal test code; all 4 cases fail with EISDIR on Linux, _none_ should open the dir successfully.
```
int f;
FILE *fp;

f = open("/tmp", O_WRONLY);
if (f >= 0) {
	puts("open tmp OK");
	close(f);
} else {
	perror("open tmp");
}

f = open("/tmp", O_WRONLY|O_CREAT);
if (f >= 0) {
	puts("open tmp CREAT OK");
	close(f);
} else {
	perror("open tmp CREAT");
}

fp = fopen("/tmp", "wb");
if (fp) {
	puts("fopen tmp OK");
	fclose(fp);
} else {
	perror("fopen tmp");
}

fp = fopen("/tmp", "a");
if (fp) {
	puts("fopen tmp a OK");
	fclose(fp);
} else {
	perror("fopen tmp a");
}
```

### CODE REVIEW

File: lib/vfscore/syscalls.c
Comment (dinhngtu) [inline lib/vfscore/syscalls.c:1]: https://github.com/unikraft/unikraft/blob/ab99b0a6de3d8ef1dfeec37dcd3423a47388b407/lib/vfscore/syscalls.c#L197 should be `goto out_unlock;`

**Review by dinhngtu — COMMENTED** [review]
Hi, I've left a comment.

File: lib/vfscore/syscalls.c
Comment (andreittr) [reply lib/vfscore/syscalls.c:1]: Oh nice, good catch, thanks! Gonna fix asap.

Comment (andreittr) [reply lib/vfscore/syscalls.c:1]: Pushed fix.

**Comment by dinhngtu** [comment]
That's all from me, thanks.

Reviewed-by: Tu Dinh Ngoc <dinhngoc.tu@irit.fr>

**Review by dinhngtu — APPROVED** [review]

**Review by razvand — APPROVED** [review]
Approved-by: Razvan Deaconescu <razvand@unikraft.io>

**Pull request closed without merging**